synthesizer/src/vm/finalize.rs

```diff
@@ -813,13 +844,45 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
         }
 
         // If the transaction is a deployment, ensure that it is not another deployment in the block from the same public fee payer.
-        if let Transaction::Deploy(_, _, _, fee) = transaction {
+        if let Transaction::Deploy(_, _, deployment, fee) = transaction {
             // If any public deployment payer has already deployed in this block, abort the transaction.
             if let Some(payer) = fee.payer() {
                 if deployment_payers.contains(&payer) {
                     return Some(format!("Another deployment in the block from the same public fee payer {payer}"));
                 }
             }
+            // Activate extra checks from CONSENSUS_V2_HEIGHT onwards.
+            if current_block_height >= N::CONSENSUS_V2_HEIGHT {
+                // Check that the synthesis cost does not exceed the maximum.
+                let Ok(synthesis_cost) = synthesis_cost(deployment) else {
+                    return Some("Failed to get synthesis cost".to_string());
+                };
+                // If the synthesis cost exceeds the block spend limit, abort the transaction.
+                if synthesis_cost.saturating_add(microcredits_spent_on_compute) > N::BLOCK_SPEND_LIMIT {
+                    return Some("Exceeds block spend limit".to_string());
+                }
+            }
+        }
+
+        // Activate extra checks from CONSENSUS_V2_HEIGHT onwards.
+        if current_block_height >= N::CONSENSUS_V2_HEIGHT {
+            // If the transaction is an execution, ensure that the total finalize cost does not exceed the block spend limit.
+            if let Transaction::Execute(_, execution, _) = transaction {
```

Comment:
nit: it might be a good idea to swap these conditions in order to not suggest that the code block below contains the only `CONSENSUS_V2` logic; alternatively, that extra logic applicable to deploy txs could be moved under this code block (without altering the order), keeping all the `V2` logic in one place

Reply:
Agree I like the clarity: fcbbb688b